File: sorting/quick_sort/quick_sort.c
// Copyright 2022 Zncl2222
#include <stdio.h>
#include <stdlib.h>
#include "quick_sort.h"

void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int partition(int* array, int front, int end) {
    int pivotkey = array[front];

    while (front < end) {
        while (front < end && array[end] >= pivotkey) {
            end--;
        }
        swap(&array[end], &array[front]);

        while (front < end && array[front] <= pivotkey) {
            front++;
        }
        swap(&array[end], &array[front]);  //  NOLINT
    }

    return front;
}

void quicksort(int* array, int front, int end) {
    int pivot;

    if (front < end) {
        pivot = partition(array, front, end);
        quicksort(array, front, pivot - 1);
        quicksort(array, pivot + 1, end);
    }
}

void print_array(int* array, int array_size) {
    for (int i = 0; i < array_size; i++) {
        printf("%d ", array[i]);
    }
    printf("\n");
}

File: sorting/golang/quick_sort.go
package sorting

func QuickSort(data []int) {
	if len(data) <= 1 {
		return
	}

	pivot := partition(data)

	QuickSort(data[:pivot])
	QuickSort(data[pivot+1:])
}

func partition(data []int) int {
	pivot := data[0]
	left, right := 1, len(data)-1

	for left <= right {
		for left <= right && data[left] <= pivot {
			left++
		}

		for left <= right && data[right] > pivot {
			right--
		}

		if left <= right {
			data[left], data[right] = data[right], data[left]
		}
	}

	data[0], data[right] = data[right], data[0]

	return right
}
